Migrate r.info to pytest

kritibirda26 opened the pull request:
Create the required sample data, pytest fixtures and migrate r.info module from gunittest to pytest.

echoix commented:
For a single fixture used by a single file, placing it in a conftest.py file, that isn't reused anywhere else, I think it is only hiding the magic pytest does (the conftest isn't directly imported by us) and readers/reviewers need to remember to go see another file. 

So, to replace a series of setUpClass/setUp, using fixtures defined in the same test file seems more efficient.

kritibirda26 commented:
The existing pytest tests are currently using per module conftest.py so I added it as a separate file in this case too. I think for now it would be fine even if we moved the fixtures to the test file itself.

But in the future, when I have more understanding of the test requirements of various modules, I would like to factor common fixtures to a higher level to reduce code duplication.

wenzeslaus commented on raster/r.info/tests/test_r_info.py:
```diff
@@ -0,0 +1,160 @@
+
+
+def assert_rinfo_output(expected_str, received_str):
+    expected = dict(**text_to_keyvalue(expected_str, sep="=", skip_empty=True))
+    received = dict(**text_to_keyvalue(received_str, sep="=", skip_empty=True))
```
Prefer JSON outputs over text_to_keyvalue.

kritibirda26 replied:
Most of the uses of this are for shell flag tests, hence I don't think JSON is the right choice here. But yes can compare strings directly.

wenzeslaus commented on raster/r.info/tests/test_r_info.py:
```diff
@@ -0,0 +1,160 @@
+        cells=10000
+        datatype=CELL
+        ncats=3"""
+    received_str = gs.read_command(
+        "r.info", map="landuse", flags="g", env=setup_maps.env
+    )
+    assert_rinfo_output(expected_str, received_str)
```
To test the flag specifically, you could also test the string output as is.

wenzeslaus commented on raster/r.info/tests/test_r_info.py:
```diff
@@ -0,0 +1,160 @@
+
+    assert sorted(list(expected.keys())) == sorted(list(received.keys()))
+
+    for key, value in expected.items():
+        assert_msg = f"{key}: expected {value}, received {received[key]}"
+        if isinstance(value, float):
+            assert value == pytest.approx(received[key]), assert_msg
+        else:
+            assert value == received[key], assert_msg
```
Approx takes a dict, and while it makes ints approx as well, two ints won't match with the default precision.

```python
>>> pytest.approx({"a": 1, "b": "abc"}) == {"a": 1, "b": "abc"}
True
>>> pytest.approx({"a": 1, "b": "abc"}) == {"a": 2, "b": "abc"}
False
```

This is one of the good ideas in pytest, you should be able to run the small things in interactive shell.

wenzeslaus commented on raster/r.info/tests/test_r_info.py:
```diff
@@ -0,0 +1,160 @@
+        assert field in received
+        received.pop(field)
+
+    assert sorted(list(expected.keys())) == sorted(list(received.keys()))
```
pytest will likely provide a good enough message without this. Always test what pytest gives you with a minimal assert statement. A lot of gunittest code is just trying to compensate for lack of specific messages or doing think because why not when we have already a long custom function anyway. (I'm the author of grass.gunittest...and, as I said elsewhere, I would do everything differently now.)

wenzeslaus commented on raster/r.info/tests/test_r_info.py:
```diff
@@ -0,0 +1,160 @@
+    # the following fields vary with the Grass sample data's path
+    # therefore only check for their presence in the JSON output
+    # and not exact values
+    remove_fields = ["location", "project", "database", "date", "creator"]
+    for field in remove_fields:
+        assert field in received
+        received.pop(field)
```
I'm not sure what to do with this. I didn't find a good way how to do this in pytest with a one-liner. `unittest.mock.ANY` seems to be usable with pytest as a dict value when pytest-mock is installed. That may be little too much for this PR, but good to keep in mind.

wenzeslaus commented on raster/r.info/tests/test_r_info.py:
```diff
@@ -0,0 +1,160 @@
+        "stddev": 0.817985,
+        "mean": 2.03,
+    }
+    output = gs.read_command(
+        "r.info", map="landuse", flags="gs", format="json", env=setup_maps.env
+    )
+    received = json.loads(output)
```
parse_command

wenzeslaus commented:
> The existing pytest tests are currently using per module conftest.py so I added it as a separate file in this case too. I think for now it would be fine even if we moved the fixtures to the test file itself.
> 
> But in the future, when I have more understanding of the test requirements of various modules, I would like to factor common fixtures to a higher level to reduce code duplication.

I agree, some fixtures can be in the files themselves which makes it more straightforward in a way, but I was putting them to separate files for two reasons: 1) the idea that they can be refactored and 2) practically, fixture in the test file itself triggers warning about shadowing symbol from the upper context (fixture is a function _and_ parameter name).
